File: stm32f103zet6传感器资料/OLED屏幕/0.96inch_SPI_OLED_Module_SSD1306_MSP096X_V1.0/1-Demo/Demo_MSP430/0.96inch_OLED_Demo_MSP430F149_Hardware_3-wire_SPI/OLED/oled.c
//////////////////////////////////////////////////////////////////////////////////	 
//本程序只供学习使用，未经作者许可，不得用于其它任何用途
//测试硬件：单片机MSP430F149,MSP430开发板,主频8MHZ，晶振8MHZ
//QDtech-TFT液晶驱动 for MSP430 硬件SPI
//xiao冯@ShenZhen QDtech co.,LTD
//公司网站:www.qdtft.com
//淘宝网站：http://qdtech.taobao.com
//wiki技术网站：http://www.lcdwiki.com
//我司提供技术支持，任何技术问题欢迎随时交流学习
//固话(传真) :+86 0755-23594567 
//手机:15989313508（冯工） 
//邮箱:lcdwiki01@gmail.com    support@lcdwiki.com    goodtft@163.com
//技术支持QQ:3002773612  3002778157
//技术交流QQ群:324828016
//创建日期:2018/8/29
//版本：V1.0
//版权所有，盗版必究。
//Copyright(C) 深圳市全动电子技术有限公司 2018-2028
//All rights reserved
/****************************************************************************************************
//=========================================电源接线================================================//
// OLED模块                C51单片机
//   VCC         接       DC 5V/3.3V      //OLED屏电源正
//   GND         接          GND          //OLED屏电源地
//=======================================液晶屏数据线接线==========================================//
//本模块默认数据总线类型为3线制SPI
// OLED模块                C51单片机
//   D1          接          P31          //OLED屏SPI写信号
//=======================================液晶屏控制线接线==========================================//
// OLED模块                C51单片机
//   CS          接          P20          //OLED屏片选控制信号
//   RES         接          P22          //OLED屏复位控制信号
//   D0          接          P33          //OLED屏SPI时钟信号
//=========================================触摸屏接线=========================================//
//本模块不带触摸功能，所以不需要触摸屏接线
****************************************************************************************************/	
/***************************************************************************************************
  * @attention
  *
  * THE PRESENT FIRMWARE WHICH IS FOR GUIDANCE ONLY AIMS AT PROVIDING CUSTOMERS
  * WITH CODING INFORMATION REGARDING THEIR PRODUCTS IN ORDER FOR THEM TO SAVE
  * TIME. AS A RESULT, QD electronic SHALL NOT BE HELD LIABLE FOR ANY
  * DIRECT, INDIRECT OR CONSEQUENTIAL DAMAGES WITH RESPECT TO ANY CLAIMS ARISING
  * FROM THE CONTENT OF SUCH FIRMWARE AND/OR THE USE MADE BY CUSTOMERS OF THE
  * CODING INFORMATION CONTAINED HEREIN IN CONNECTION WITH THEIR PRODUCTS.
****************************************************************************************************/	
#include "oled.h"
#include "delay.h"
#include "stdlib.h"
#include "string.h" 	 
#include "spi.h"

//OLED显存总共分为8页
//每页8行，一行128个像素点
//OLED的显存
//存放格式如下.
//[0]0 1 2 3 ... 127 (0~7)行	   
//[1]0 1 2 3 ... 127 (8~15)行	
//[2]0 1 2 3 ... 127 (16~23)行	
//[3]0 1 2 3 ... 127 (24~31)行	
//[4]0 1 2 3 ... 127 (32~39)行	
//[5]0 1 2 3 ... 127 (40~47)行	
//[6]0 1 2 3 ... 127 (48~55)行	
//[7]0 1 2 3 ... 127 (56~63)行			   

/*******************************************************************
 * @name       :void OLED_WR_Byte(unsigned dat,unsigned cmd)
 * @date       :2018-08-27
 * @function   :Write a byte of content to the OLED screen
 * @parameters :dat:Content to be written
                cmd:0-write command
								    1-write data
 * @retvalue   :None
********************************************************************/
void OLED_WR_Byte(unsigned dat,unsigned cmd)
{
        u16 data=0;
        data=((cmd<<15)|(dat<<7));
	OLED_CS_Clr;
	SPI_WriteByte((data>>8)&0xFF);
        SPI_WriteByte(data&0xFF);
	OLED_CS_Set;
}

/*******************************************************************
 * @name       :void OLED_Set_Pos(unsigned char x, unsigned char y) 
 * @date       :2018-08-27
 * @function   :Set coordinates in the OLED screen
 * @parameters :x:x coordinates
                y:y coordinates
 * @retvalue   :None
********************************************************************/
void OLED_Set_Pos(unsigned char x, unsigned char y) 
{
 	OLED_WR_Byte(YLevel+y,OLED_CMD);
	OLED_WR_Byte(((x&0xf0)>>4)|0x10,OLED_CMD);
	OLED_WR_Byte((x&0x0f),OLED_CMD); 
}  
   			
/*******************************************************************
 * @name       :void OLED_Clear(unsigned dat)  
 * @date       :2018-08-27
 * @function   :clear OLED screen
 * @parameters :dat:0-Display full black
                    1-Display full white
 * @retvalue   :None
********************************************************************/ 
void OLED_Clear(unsigned dat)  
{ 
	u8 i,n,color;	
	if(dat)
	{
		color = 0xff;
	}
	else
	{
		color = 0;
	}
	for(i=0;i<PAGE_SIZE;i++)  
	{  
		OLED_WR_Byte(YLevel+i,OLED_CMD);    //设置页地址（0~7）
		OLED_WR_Byte(XLevelL,OLED_CMD);      //设置显示位置—列低地址
		OLED_WR_Byte(XLevelH,OLED_CMD);      //设置显示位置—列高地址   
		for(n=0;n<WIDTH;n++)
		{
			OLED_WR_Byte(color,OLED_DATA); 
		}
	}   //更新显示
}

/*******************************************************************
 * @name       :void OLED_Reset(void) 
 * @date       :2018-08-27
 * @function   :Reset OLED screen
 * @parameters :dat:0-Display full black
                    1-Display full white
 * @retvalue   :None
********************************************************************/ 
void OLED_Reset(void)
{
	OLED_RST_Set;
	delay_ms(100);
	OLED_RST_Clr;
	delay_ms(100);
	OLED_RST_Set; 
}	

/*******************************************************************
 * @name       :void OLED_GPIO_Init(void)
 * @date       :2018-08-27
 * @function   :initialise OLED non SPI GPIO
 * @parameters :None
 * @retvalue   :None
********************************************************************/
void OLED_GPIO_Init(void)
{
  P2DIR |= OLED_CS+OLED_RST;
}
/*******************************************************************
 * @name       :void OLED_Init(void)
 * @date       :2018-08-27
 * @function   :initialise OLED SH1106 control IC
 * @parameters :None
 * @retvalue   :None
********************************************************************/ 				    
void OLED_Init(void)
{
	SPI_Init();  //初始化SPI
        OLED_GPIO_Init(); //初始化GPIO
 	delay_ms(200);
	OLED_Reset();     //复位OLED

/**************初始化SSD1306*****************/	
	OLED_WR_Byte(0xAE,OLED_CMD); /*display off*/
	OLED_WR_Byte(0x00,OLED_CMD); /*set lower column address*/
	OLED_WR_Byte(0x10,OLED_CMD); /*set higher column address*/
	OLED_WR_Byte(0x40,OLED_CMD); /*set display start line*/ 
	OLED_WR_Byte(0xB0,OLED_CMD); /*set page address*/
	OLED_WR_Byte(0x81,OLED_CMD); /*contract control*/ 
	OLED_WR_Byte(0xFF,OLED_CMD); /*128*/
	OLED_WR_Byte(0xA1,OLED_CMD); /*set segment remap*/ 
	OLED_WR_Byte(0xA6,OLED_CMD); /*normal / reverse*/
	OLED_WR_Byte(0xA8,OLED_CMD); /*multiplex ratio*/ 
	OLED_WR_Byte(0x3F,OLED_CMD); /*duty = 1/64*/
	OLED_WR_Byte(0xC8,OLED_CMD); /*Com scan direction*/
	OLED_WR_Byte(0xD3,OLED_CMD); /*set display offset*/ 
	OLED_WR_Byte(0x00,OLED_CMD);
	OLED_WR_Byte(0xD5,OLED_CMD); /*set osc division*/ 
	OLED_WR_Byte(0x80,OLED_CMD);
	OLED_WR_Byte(0xD9,OLED_CMD); /*set pre-charge period*/ 
	OLED_WR_Byte(0XF1,OLED_CMD);
	OLED_WR_Byte(0xDA,OLED_CMD); /*set COM pins*/ 
	OLED_WR_Byte(0x12,OLED_CMD);
	OLED_WR_Byte(0xDB,OLED_CMD); /*set vcomh*/ 
	OLED_WR_Byte(0x30,OLED_CMD);
	OLED_WR_Byte(0x8D,OLED_CMD); /*set charge pump disable*/ 
	OLED_WR_Byte(0x14,OLED_CMD);
	OLED_WR_Byte(0xAF,OLED_CMD); /*display ON*/
}  

File: 传感器资料/OLED屏幕/0.96inch_SPI_OLED_Module_SSD1306_MSP096X_V1.0/1-Demo/Demo_RaspberryPI/0.96inch_OLED_Demo_python_Hardware_4-wire_SPI/source/oled.py
# This program is for learning only,Not to be used for any other purpose 
# without the permission of the author
# Testing Hardware:Raspberry PI 3 ,  basic frequency:1GHz
# QDtech-OLED liquid crystal driver for Raspberry PI
# xiaofeng@ShenZhen QDtech co.,LTD
# Company Website:www.qdtft.com
# Taobao Website:http://qdtech.taobao.com
# wiki Technology Website:http://www.lcdwiki.com
# We provide technical support,Any technical questions are welcome to 
# exchange and study at any time
# Fixed telephone (fax):+86 0755-23594567 
# cell-phone:15989313508(Mr Feng)
# E-mail:lcdwiki01@gmail.com    support@lcdwiki.com    goodtft@163.com
# Technical Support QQ:3002773612  3002778157
# Technical Exchange QQ group:324828016
# Date:2018/9/6
# Version:V1.0
# Copyright reserved, piracy must be investigated
# Copyright(C) ShenZhen QDtech co.,LTD 2018-2028
# All rights reserved
#****************************************************************************************************
# This program uses the bcm2835 gpio library,
# so the pin definition using BCM coding
#=====================================power supply wiring===========================================//
# OLED Module                Raspberry PI    
#    VCC        connect       DC 5V/3.3V      //OLED power positive, Physical pin 1,2,4
#    GND        connect          GND          //OLED power ground,Physical pin 6,9,14,20,25,30,34,39
#======================================data line wiring=============================================//
# The default data bus type for this module is 4-wire SPI
#  OLED Module                Raspberry PI 
#    D1         connect       19(bcm:10)      //OLED spi write signal
#======================================control line wiring==========================================//
# OLED Module                Raspberry PI 
#    CS         connect       24(bcm:8)       //OLED chip selection control signal
#    RES        connect        5(bcm:3)       //OLED reset control signal
#    DC         connect        3(bcm:2)       //OLED data or command selection control signal
#    D0         connect       23(bcm:11)      //OLED spi colck signal
#========================================touch screen wiring========================================//
#This module has no touch function,so it don't need touch screen wiring
#*****************************************************************************************************/	
#****************************************************************************************************
# @attention
#
# THE PRESENT FIRMWARE WHICH IS FOR GUIDANCE ONLY AIMS AT PROVIDING CUSTOMERS
# WITH CODING INFORMATION REGARDING THEIR PRODUCTS IN ORDER FOR THEM TO SAVE
# TIME. AS A RESULT, QD electronic SHALL NOT BE HELD LIABLE FOR ANY
# DIRECT, INDIRECT OR CONSEQUENTIAL DAMAGES WITH RESPECT TO ANY CLAIMS ARISING
# FROM THE CONTENT OF SUCH FIRMWARE AND/OR THE USE MADE BY CUSTOMERS OF THE
# CODING INFORMATION CONTAINED HEREIN IN CONNECTION WITH THEIR PRODUCTS.
#*****************************************************************************************************/

import spidev
import RPi.GPIO as GPIO
import time

OLED_COMMAND = 0
OLED_DATA = 1

class SSD1306(object):
	"""class for SSD1306 64*128 1.3inch SPI OLED module."""

	def __init__(self,cs,res,dc,spi):
		# set oled display parameter 
		self.width = 128
		self.height = 64
		self.pagesize = 8
		self.ylevel = 0xB0
		self.xlevell = 0x00
		self.xlevelh = 0x10
		self.oledbuffer = [0]*(self.width*self.pagesize)
		# Initialize oled pin
		self.oledcs = cs
		self.oledres = res
		self.oleddc = dc
		GPIO.setmode(GPIO.BCM)
		GPIO.setwarnings(False)
		GPIO.setup(self.oledcs,GPIO.OUT)
		GPIO.setup(self.oledres,GPIO.OUT)
		GPIO.setup(self.oleddc,GPIO.OUT)
		# Initialize spi
		self.oledspi = spi
	def writebyte(self,val,flag):
		"""send one byte data to oled module"""
		if flag == OLED_COMMAND:
			GPIO.output(self.oleddc,GPIO.LOW)
		else:
			GPIO.output(self.oleddc,GPIO.HIGH)
		GPIO.output(self.oledcs,GPIO.LOW)
		self.oledspi.writebytes([val])
		self.oledspi.xfer([val],8000000)
		GPIO.output(self.oledcs,GPIO.HIGH)
	def oledreset(self):
		"""reset the oled module"""
		GPIO.output(self.oledres,GPIO.HIGH)
		time.sleep(0.01)
		GPIO.output(self.oledres,GPIO.LOW)
		time.sleep(0.01)
		GPIO.output(self.oledres,GPIO.HIGH)
	def oledinit(self):
		self.oledreset()
		"""init SSD1306"""	
		self.writebyte(0xAE,OLED_COMMAND); #display off
		self.writebyte(0x00,OLED_COMMAND); #set lower column address
		self.writebyte(0x10,OLED_COMMAND); #set higher column address
		self.writebyte(0x40,OLED_COMMAND); #set display start line
		self.writebyte(0xB0,OLED_COMMAND); #set page address
		self.writebyte(0x81,OLED_COMMAND); #contract control
		self.writebyte(0xFF,OLED_COMMAND); #128
		self.writebyte(0xA1,OLED_COMMAND); #set segment remap 
		self.writebyte(0xA6,OLED_COMMAND); #normal / reverse
		self.writebyte(0xA8,OLED_COMMAND); #multiplex ratio 
		self.writebyte(0x3F,OLED_COMMAND); #duty = 1/64
		self.writebyte(0xC8,OLED_COMMAND); #Com scan direction
		self.writebyte(0xD3,OLED_COMMAND); #set display offset 
		self.writebyte(0x00,OLED_COMMAND);
		self.writebyte(0xD5,OLED_COMMAND); #set osc division
		self.writebyte(0x80,OLED_COMMAND);
		self.writebyte(0xD9,OLED_COMMAND); #set pre-charge period
		self.writebyte(0XF1,OLED_COMMAND);
		self.writebyte(0xDA,OLED_COMMAND); #set COM pins 
		self.writebyte(0x12,OLED_COMMAND);
		self.writebyte(0xDB,OLED_COMMAND); #set vcomh 
		self.writebyte(0x30,OLED_COMMAND);
		self.writebyte(0x8D,OLED_COMMAND); #set charge pump disable 
		self.writebyte(0x14,OLED_COMMAND);
		self.writebyte(0xAF,OLED_COMMAND); #display ON
	def oledclear(self,val):
		"""clear oled image buffer"""
		if val == 1:
			self.oledbuffer = [0xff]*(self.width*self.pagesize)
		else:
			self.oledbuffer = [0]*(self.width*self.pagesize)
	def oleddisplay(self):
		"""send oled image buffer data to oled module"""
		for i in range(self.pagesize):
			self.writebyte(self.ylevel+i,OLED_COMMAND)
			self.writebyte(self.xlevell,OLED_COMMAND)
			self.writebyte(self.xlevelh,OLED_COMMAND)
			for n in range(self.width):
				self.writebyte(self.oledbuffer[i*self.width+n],OLED_DATA)
	def oledinvertdisplay(self,val):
		if val == 1:
			self.writebyte(0xA7,OLED_COMMAND)
		else:
			self.writebyte(0xA6,OLED_COMMAND)
	def oledimage(self,image):
		"""set the value of Python Image Library to oled buffer"""
		if image.mode != '1':
			raise ValueError('Image must be in mode 1.')
		imgwidth,imgheight = image.size
		if imgwidth != self.width or imgheight != self.height:
			raise ValueError('Image must be same dimensions as display({0}x{1}).' .format(self.width, self.height))
		pixel = image.load()
		for i in range(self.pagesize):
			for j in range(self.width):
				val = 0
				for k in [0,1,2,3,4,5,6,7]:
                                        val = val << 1
				        val |=0 if pixel[(j,i*8+7-k)] == 0 else 1
				self.oledbuffer[i*self.width+j] = val 
